diff -aurN a/iotedge/src/config/apply.rs b/iotedge/src/config/apply.rs
--- a/iotedge/src/config/apply.rs	2024-05-02 10:41:51.470870089 +0200
+++ b/iotedge/src/config/apply.rs	2024-05-02 10:49:00.046186818 +0200
@@ -3,7 +3,7 @@
 //! This subcommand takes the super-config file, converts it into the individual services' config files,
 //! writes those files, and restarts the services.
 
-use std::{collections::HashMap, path::Path};
+use std::{collections::HashMap, path::Path, env};
 
 use aziotctl_common::config as common_config;
 
@@ -181,8 +181,12 @@
 
     println!("Azure IoT Edge has been configured successfully!");
     println!();
-    println!("Restarting service for configuration to take effect...");
-    crate::System::system_restart().map_err(|err| format!("{err}"))?;
+    if env::var("OMIT_SERVICE_RESTART").is_ok() {
+        println!("Skipping service restart!"); 
+    } else {
+        println!("Restarting service for configuration to take effect...");
+        crate::System::system_restart().map_err(|err| format!("{err}"))?;
+    }
     println!("Done.");
 
     Ok(())
