diff -aurN a/iotedge/src/check/mod.rs b/iotedge/src/check/mod.rs
--- a/iotedge/src/check/mod.rs	2021-04-13 08:11:18.029260302 +0200
+++ b/iotedge/src/check/mod.rs	2021-04-13 19:19:41.789235353 +0200
@@ -237,7 +237,7 @@
             description: String,
             result: CheckResult,
             additional_info: serde_json::Value,
-        };
+        }
 
         let mut checks: BTreeMap<String, CheckOutputSerializable> = Default::default();
 
