diff -aurN a/iotedge/src/check/mod.rs b/iotedge/src/check/mod.rs
--- a/iotedge/src/check/mod.rs	2021-04-13 08:11:18.029260302 +0200
+++ b/iotedge/src/check/mod.rs	2021-04-13 19:19:41.789235353 +0200
@@ -237,7 +237,7 @@
             description: String,
             result: CheckResult,
             additional_info: serde_json::Value,
-        };
+        }
 
         let mut checks: BTreeMap<String, CheckOutputSerializable> = Default::default();
 
diff --git a/iotedge/src/config/import/mod.rs b/iotedge/src/config/import/mod.rs
index 4a431b7d0..ef0abcf15 100644
--- a/iotedge/src/config/import/mod.rs
+++ b/iotedge/src/config/import/mod.rs
@@ -270,7 +270,7 @@ fn execute_inner(
                         id_scope: scope_id,
                         attestation: common_config::super_config::DpsAttestationMethod::X509 {
                             // TODO: Remove this when IS supports registration ID being optional for DPS-X509
-                            registration_id: registration_id
+                            registration_id: Some(registration_id)
                                 .ok_or_else(|| "registration ID is currently required")?,
                             identity: common_config::super_config::X509Identity::Preloaded {
                                 identity_cert,
