diff -aurN a/pkg/networkmanager/firewall.jsx b/pkg/networkmanager/firewall.jsx
--- a/pkg/networkmanager/firewall.jsx	2024-06-06 13:10:13.861684039 +0200
+++ b/pkg/networkmanager/firewall.jsx	2024-06-06 14:21:24.020326679 +0200
@@ -762,11 +762,7 @@
 
     save(event) {
         const Dialogs = this.context;
-        let p;
-        if (firewall.zones[this.state.zone].services.indexOf("cockpit") === -1)
-            p = firewall.addService(this.state.zone, "cockpit");
-        else
-            p = Promise.resolve();
+        const p = Promise.resolve();
 
         const sources = this.state.ipRange === "ip-range" ? this.state.ipRangeValue.split(",").map(ip => ip.trim()) : [];
         p.then(() =>
@@ -949,11 +945,7 @@
 
     onRemoveZone(zone) {
         const Dialogs = this.context;
-        let body;
-        if (firewall.zones[zone].services.indexOf("cockpit") !== -1)
-            body = _("This zone contains the cockpit service. Make sure that this zone does not apply to your current web console connection.");
-        else
-            body = _("Removing the zone will remove all services within it.");
+        const body = _("Removing the zone will remove all services within it.");
         Dialogs.show(<DeleteConfirmationModal title={ cockpit.format(_("Remove zone $0"), zone) }
                                               body={body}
                                               target={zone}
