diff --git a/configure.ac b/configure.ac
index 76bdcae6f..3611ecddd 100644
--- a/configure.ac
+++ b/configure.ac
@@ -16,7 +16,7 @@
 # along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
 
 AC_INIT([Cockpit],
-        [m4_esyscmd([tools/git-version-gen .tarball])],
+        [m4_esyscmd_s([if [ -f .tarball ]; then cat .tarball; else echo $(tools/git-version-gen); fi])],
         [devel@lists.cockpit-project.org],
         [cockpit],
         [https://cockpit-project.org/])
