diff -aurN a/msg_sink_mod/msg_sink_mod.c b/msg_sink_mod/msg_sink_mod.c
--- a/msg_sink_mod/msg_sink_mod.c	2019-04-25 14:41:07.479911512 +0200
+++ b/msg_sink_mod/msg_sink_mod.c	2019-04-25 14:56:19.944465406 +0200
@@ -716,15 +716,18 @@
     uint32_t ulByteIdx;
     uint8_t  bByte;
 
-    snprintf(pszTextBuf, ulTextBufSize, "\"");
+    *pszTextBuf++='\"';
 
     for (ulByteIdx = 0; ulByteIdx < ulVarSize; ulByteIdx++)
     {
+      const char hexmap[] = {'0', '1', '2', '3', '4', '5', '6', '7',
+                             '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
       bByte = *(pbValue + ulByteIdx);
-      snprintf(pszTextBuf + 1 + (ulByteIdx * 2), ulTextBufSize - 1 - (ulByteIdx * 2), "%02x", bByte);
+      *pszTextBuf++ = hexmap[(bByte & 0xF0) >> 4];
+      *pszTextBuf++ = hexmap[bByte & 0xF0];
     }
 
-    snprintf(pszTextBuf + 1 + (ulByteIdx * 2), ulTextBufSize - 1 - (ulByteIdx * 2), "\"");
+    *pszTextBuf='\"';
   }
   break;
 
