diff -aurN a/msg_sink_mod/msg_sink_mod.c b/msg_sink_mod/msg_sink_mod.c
--- a/msg_sink_mod/msg_sink_mod.c	2019-04-25 07:47:31.000000000 +0200
+++ b/msg_sink_mod/msg_sink_mod.c	2019-04-25 10:54:48.182149586 +0200
@@ -13,194 +13,6 @@
 #define VERBOSE_PRINT(...)
 #endif
 
-typedef struct MESSAGE_INSTANCE_TAG
-{
-    IOTHUB_MESSAGE_HANDLE messageHandle;
-    size_t messageTrackingId;  // For tracking the messages within the user callback.
-    MODULE_INSTANCE_T* ptInstance;
-} 
-MESSAGE_INSTANCE;
-
-uint64_t ullErrorCnt = 0;
-uint64_t ullSendOk = 0;
-uint64_t ullAckOk = 0;
-uint64_t ullAckNok = 0;
-
-
-// SendConfirmationCallback is invoked when the message that was forwarded on from 'InputQueue1Callback'
-// pipeline function is confirmed.
-static void SendConfirmationCallback(IOTHUB_CLIENT_CONFIRMATION_RESULT result, void* userContextCallback)
-{
-    // The context corresponds to which message# we were at when we sent.
-    MESSAGE_INSTANCE* messageInstance = (MESSAGE_INSTANCE*)userContextCallback;
-    if (messageInstance == NULL || messageInstance->ptInstance == NULL || messageInstance->ptInstance->tIotConfig.fIOTLibLoadOk == 0 )
-    {
-      // TODO: error handling 
-      ullAckNok++;
-      return; 
-    }
-    IOT_CONFIG_T* ptIoTCfg = &messageInstance->ptInstance->tIotConfig;
-    ullAckOk++;
-    VERBOSE_PRINT("Confirmation[%zu] received for message with result = %d\r\n", messageInstance->messageTrackingId, result);
-    if (ptIoTCfg != NULL)
-      ptIoTCfg->pfnIoTHubMessage_Destroy(messageInstance->messageHandle);
-    free(messageInstance);
-}
-
-// Allocates a context for callback and clones the message
-// NOTE: The message MUST be cloned at this stage.  InputQueue1Callback's caller always frees the message
-// so we need to pass down a new copy.
-/* TODO: this needs to be adjusted for the variable info_t  */
-static MESSAGE_INSTANCE* CreateMessageInstance(char* szMessage, MODULE_INSTANCE_T* ptInstance)
-{
-    MESSAGE_INSTANCE* messageInstance = (MESSAGE_INSTANCE*)malloc(sizeof(MESSAGE_INSTANCE));
-    if (NULL == messageInstance || ptInstance == NULL )
-    {
-        printf("Failed allocating 'MESSAGE_INSTANCE' for pipelined message\r\n");
-    }
-    else
-    {
-        memset(messageInstance, 0, sizeof(*messageInstance));
-
-        if ((messageInstance->messageHandle = ptInstance->tIotConfig.pfnIoTHubMessage_CreateFromString(szMessage)) == NULL)
-        {
-            free(messageInstance);
-            messageInstance = NULL;
-            ullErrorCnt++;
-        }
-        else
-        {
-            //messageInstance->messageTrackingId = messagesReceivedByInput1Queue;
-            messageInstance->ptInstance = ptInstance;
-            messageInstance->messageTrackingId = ptInstance->tIotConfig.ullMessagesReceivedByInput1Queue;
-            //MAP_HANDLE propMap = IoTHubMessage_Properties(messageInstance->messageHandle);
-            //if (Map_AddOrUpdate(propMap, "MessageType", "Alert") != MAP_OK)
-            //{
-            //    printf("ERROR: Map_AddOrUpdate Failed!\r\n");
-            //}
-        }
-    }
-    return messageInstance;
-}
-
-/* This function creates the client handle from environment */
-IOTHUB_MODULE_CLIENT_LL_HANDLE InitializeConnection( MODULE_INSTANCE_T* ptInstance)
-{
-    IOTHUB_MODULE_CLIENT_LL_HANDLE iotHubModuleClientHandle;
-    // TODO: error handling input checking
-    if (ptInstance->tIotConfig.pfnIoTHub_Init() != 0)
-    {
-        printf("Failed to initialize the platform.\r\n");
-        iotHubModuleClientHandle = NULL;
-    } 
-    else if ((iotHubModuleClientHandle = ptInstance->tIotConfig.pfnIoTHubModuleClient_LL_CreateFromEnvironment(ptInstance->tIotConfig.pfnMQTT_Protocol)) == NULL)
-    {
-        printf("ERROR: IoTHubModuleClient_LL_CreateFromEnvironment failed\r\n");
-    }
-    else
-    {
-        // Uncomment the following lines to enable verbose logging.
-        // bool traceOn = true;
-        // IoTHubModuleClient_LL_SetOption(iotHubModuleClientHandle, OPTION_LOG_TRACE, &trace);
-    }
-
-    return iotHubModuleClientHandle;
-}
-
-/* this function will delete the iothub client handle */
-static void DeInitializeConnection(IOTHUB_MODULE_CLIENT_LL_HANDLE iotHubModuleClientHandle,  MODULE_INSTANCE_T* ptInstance )
-{
-  if (ptInstance != NULL)
-  {
-    if (iotHubModuleClientHandle != NULL)
-    {
-        ptInstance->tIotConfig.pfnIoTHubModuleClient_LL_Destroy(iotHubModuleClientHandle);
-    }
-    ptInstance->tIotConfig.pfnIoTHub_Deinit();
-  }
-}
-
-/* This function creates the client connection and initializes the worker thread and mutex */
-void iothub_module_deinit(void* pvDestinationInstanceHandle)
-{
-  MODULE_INSTANCE_T* ptInstance = (MODULE_INSTANCE_T*)pvDestinationInstanceHandle;
-  IOT_CONFIG_T* ptParam = &ptInstance->tIotConfig;
-  IOTHUB_MODULE_CLIENT_LL_HANDLE iotHubModuleClientHandle = ptParam->iotHubModuleClientHandle;
-
-  pthread_mutex_lock(&ptParam->tThreadLock);
-  ptParam->fInitialized = 0;   
-  pthread_mutex_unlock(&ptParam->tThreadLock);
-
-  pthread_join(ptParam->hWorkerThread, NULL);
-  ptParam->hWorkerThread = 0;
-
-  pthread_mutex_destroy(&ptParam->tThreadLock);
-
-  ptParam->iotHubModuleClientHandle = NULL;
-  DeInitializeConnection(iotHubModuleClientHandle, ptInstance);
-}
-
-
-void* iothubWorkerThread(void* pvThreadParam)
-{
-  /* The receiver just loops constantly waiting for messages. */
-  // MODULE_INSTANCE_T*             ptInstance = (MODULE_INSTANCE_T*)pvDestinationInstanceHandle;
-  int                            fWorkerInitialized;
-  IOT_CONFIG_T*                  ptParam = (IOT_CONFIG_T*)pvThreadParam;
-  IOTHUB_MODULE_CLIENT_LL_HANDLE iotHubModuleClientHandle  = ptParam->iotHubModuleClientHandle;
-  
-  struct timespec req = {0};
-  memset(&req,0,sizeof(struct timespec));
-  req.tv_nsec = 1000 * 1000 * 100; // 100 ms
-  printf("Worker thread waiting for incoming messages.\r\n");
-
-  uint32_t ulCnt = 0;
-  do 
-  {
-    pthread_mutex_lock(&ptParam->tThreadLock);
-    fWorkerInitialized = ptParam->fInitialized;
-    pthread_mutex_unlock(&ptParam->tThreadLock);
-    
-    /* All IoTHubClient interactions (in regards to network traffic 
-        and/or user level callbacks) are the effect of calling this  
-        function and they take place synchronously inside _DoWork. */
-    ptParam->pfnIoTHubModuleClient_LL_DoWork(iotHubModuleClientHandle);  
-    if (ulCnt % 50 == 0)
-    {
-      printf("sent %"PRIu64"\r\n", ptParam->ullMessagesReceivedByInput1Queue );
-      printf("ullErrorCnt: %"PRIu64" \r\n", ullErrorCnt );
-      printf("ullSendOk: %"PRIu64" \r\n", ullSendOk );
-      printf("ullAckOk: %"PRIu64" \r\n", ullAckOk );
-      printf("ullAckNok: %"PRIu64" \r\n", ullAckNok );
-
-      fflush(stdout);
-    }
-    nanosleep(&req,NULL);
-    ulCnt++;
-
-  } while (fWorkerInitialized);
-}
-
-/* This function creates the client connection and initializes the worker thread and mutex */
-void iothub_module_init(MODULE_INSTANCE_T* ptInstance)
-{
-  IOTHUB_MODULE_CLIENT_LL_HANDLE iotHubModuleClientHandle;
-  IOT_CONFIG_T* ptParam = &ptInstance->tIotConfig;
-
-  if ((iotHubModuleClientHandle = InitializeConnection(ptInstance)) != NULL )
-  {
-    /* successfully initialyzed connection; normally we would "&& SetupCallbacksForModule(iotHubModuleClientHandle) == 0"
-      configure input callbacks but we have analyzer as value source  */
-    printf("Create iothub worker thread.\r\n"); 
-    ptParam->iotHubModuleClientHandle = iotHubModuleClientHandle;
-    ptParam->fInitialized = 1;
-    pthread_mutex_init(&ptParam->tThreadLock, NULL);
-    /* this needs to be in a seperate thread only started here */
-    pthread_create(&ptParam->hWorkerThread, NULL, iothubWorkerThread, (void*)ptParam);
-  }
-}
-
-
 /*! Free internal config
 * Clears the internal configuration lists and free used memory
 *
@@ -300,6 +112,11 @@
                       {
                         ptMessageSinkParams->szAmqpTopic = strdup(json_string_value(ptAmqpTopic));
                       }
+                      /* use ampq topic for mqtt if not explicitely set */
+                      if ((NULL != ptAmqpTopic) && (NULL == ptMqttTopic))
+                      {
+                        ptMessageSinkParams->szMqttTopic = strdup(json_string_value(ptAmqpTopic));
+                      }
                       if (NULL != ptIdentifier)
                       {
                         ptMessageSinkParams->szIdentifier = strdup(json_string_value(ptIdentifier));
@@ -466,69 +283,17 @@
     ptMqttCfg = json_object_get(ptCmdObj, "mqtt");
     ptAmqpCfg = json_object_get(ptCmdObj, "amqp");
 
-    if (json_is_object(ptMqttCfg))
+    if (ptThisInst->tMqttCfg.fPahoOk == 1)
     {
-      if (ptThisInst->tMqttCfg.fPahoOk == 1)
-      {
-        json_t*  ptServerUri;
-        json_t*  ptClientId;
-
-        ptServerUri = json_object_get(ptMqttCfg, "serverUri");
-        ptClientId = json_object_get(ptMqttCfg, "clientId");
-
-        /* !!!! TODO: add more MQTT config fields from MQTTClient_connectOptions to be configured via JSON */
-
-        if (json_is_string(ptServerUri) && json_is_string(ptClientId))
-        {
-          ptThisInst->tMqttCfg.szServerUri = strdup(json_string_value(ptServerUri));
-          ptThisInst->tMqttCfg.szClientId = strdup(json_string_value(ptClientId));
-          ptThisInst->tMqttCfg.fEnabled = 1;
-        }
-        else
-        {
-          assert(0);
-          ulRes = RES_INVALID_JSON;
-        }
-      }
-      else
-      {
-        /* Paho not available, but configuration was set, return an error that MQTT is not usable */
-        assert(0);
-        ulRes = MSG_RES_NOT_AVAILABLE;
-      }
+      /* Fix to local broker */
+      ptThisInst->tMqttCfg.szServerUri = "tcp://localhost:1883";
+      ptThisInst->tMqttCfg.szClientId = "msg_sink_mod";
+      ptThisInst->tMqttCfg.fEnabled = 1;
     }else 
     {
       /* no mqtt configuration is set */
     }
 
-    if (json_is_object(ptAmqpCfg))
-    {
-      if (ptThisInst->tIotConfig.fIOTLibLoadOk == 1)
-      {
-        json_t*  ptJStatus;
-        json_t*  ptClientId;
-
-        ptJStatus = json_object_get(ptAmqpCfg, "status");
-    
-        /* !!!! TODO: add more AMQP config fields to be configured via JSON */
-        if (json_is_string(ptJStatus))
-        {
-          const char* szStatus = json_string_value(ptJStatus);
-          if (NULL != szStatus && 0 == strcmp(szStatus, "enabled"))
-          {
-            ptThisInst->tIotConfig.fEnabled = 1;
-          } else { /* default already initialyzed */ }
-        } else { 
-          /* not found or not a string */ 
-          assert(0);
-          ulRes = RES_INVALID_JSON;
-        }
-      } else { 
-        /* no az-iot-sdk-c lib loaded */        
-        assert(0);
-        ulRes = MSG_RES_NOT_AVAILABLE; 
-      }
-    } else { /* json amqp not found or not of object type */ }
   } else { /* json comand not of type object */ }
 
   /* if configuration is correct save it so it can be read out later, otherwise leave it empty */
@@ -675,54 +440,6 @@
 
           LIST_INIT(&ptThisInst->tVarList);
 
-          ptThisInst->tIotConfig.hIoTHubDll = (HMODULE)dlopen("libiothub_client.so", RTLD_NOW);      /* !!! TODO: ensure correct path for DLL */
-          if (ptThisInst->tIotConfig.hIoTHubDll  != NULL)
-          {
-            /* get all required exposed functions */
-            ptThisInst->tIotConfig.pfnIoTHub_Init                                   = (PFN_IoTHub_Init                                 )dlsym(ptThisInst->tIotConfig.hIoTHubDll, "IoTHub_Init");                            
-            ptThisInst->tIotConfig.pfnIoTHub_Deinit                                 = (PFN_IoTHub_Deinit                               )dlsym(ptThisInst->tIotConfig.hIoTHubDll, "IoTHub_Deinit");                                 
-            ptThisInst->tIotConfig.pfnIoTHubMessage_CreateFromString                = (PFN_IoTHubMessage_CreateFromString              )dlsym(ptThisInst->tIotConfig.hIoTHubDll, "IoTHubMessage_CreateFromString");                               
-            ptThisInst->tIotConfig.pfnIoTHubMessage_Destroy                         = (PFN_IoTHubMessage_Destroy                       )dlsym(ptThisInst->tIotConfig.hIoTHubDll, "IoTHubMessage_Destroy");                             
-            ptThisInst->tIotConfig.pfnIoTHubModuleClient_LL_CreateFromEnvironment   = (PFN_IoTHubModuleClient_LL_CreateFromEnvironment )dlsym(ptThisInst->tIotConfig.hIoTHubDll, "IoTHubModuleClient_LL_CreateFromEnvironment");                                
-            ptThisInst->tIotConfig.pfnIoTHubModuleClient_LL_Destroy                 = (PFN_IoTHubModuleClient_LL_Destroy               )dlsym(ptThisInst->tIotConfig.hIoTHubDll, "IoTHubModuleClient_LL_Destroy");                                
-            ptThisInst->tIotConfig.pfnIoTHubModuleClient_LL_DoWork                  = (PFN_IoTHubModuleClient_LL_DoWork                )dlsym(ptThisInst->tIotConfig.hIoTHubDll, "IoTHubModuleClient_LL_DoWork");                                 
-            ptThisInst->tIotConfig.pfnIoTHubModuleClient_LL_SendEventToOutputAsync  = (PFN_IoTHubModuleClient_LL_SendEventToOutputAsync)dlsym(ptThisInst->tIotConfig.hIoTHubDll, "IoTHubModuleClient_LL_SendEventToOutputAsync");                                 
-            ptThisInst->tIotConfig.pfnMQTT_Protocol                                 = (PFN_MQTT_Protocol                               )dlsym(ptThisInst->tIotConfig.hIoTHubDll, "MQTT_Protocol");                                 
-            
-
-            if (( ptThisInst->tIotConfig.pfnIoTHub_Init                                     == NULL ) ||
-                ( ptThisInst->tIotConfig.pfnIoTHub_Deinit                                   == NULL ) ||                  
-                ( ptThisInst->tIotConfig.pfnIoTHubMessage_CreateFromString                  == NULL ) ||                  
-                ( ptThisInst->tIotConfig.pfnIoTHubMessage_Destroy                           == NULL ) ||
-                ( ptThisInst->tIotConfig.pfnIoTHubModuleClient_LL_CreateFromEnvironment     == NULL ) ||
-                ( ptThisInst->tIotConfig.pfnIoTHubModuleClient_LL_Destroy                   == NULL ) ||
-                ( ptThisInst->tIotConfig.pfnIoTHubModuleClient_LL_DoWork                    == NULL ) ||            
-                ( ptThisInst->tIotConfig.pfnIoTHubModuleClient_LL_SendEventToOutputAsync    == NULL ) || 
-                ( ptThisInst->tIotConfig.pfnMQTT_Protocol                                   == NULL ) )                            
-            {
-              /* azure iot not fully usable, disable completely */
-              ptThisInst->tIotConfig.pfnIoTHub_Init                                  = NULL;
-              ptThisInst->tIotConfig.pfnIoTHub_Deinit                                = NULL;
-              ptThisInst->tIotConfig.pfnIoTHubMessage_CreateFromString               = NULL;
-              ptThisInst->tIotConfig.pfnIoTHubMessage_Destroy                        = NULL;
-              ptThisInst->tIotConfig.pfnIoTHubModuleClient_LL_CreateFromEnvironment  = NULL;
-              ptThisInst->tIotConfig.pfnIoTHubModuleClient_LL_Destroy                = NULL;
-              ptThisInst->tIotConfig.pfnIoTHubModuleClient_LL_DoWork                 = NULL;
-              ptThisInst->tIotConfig.pfnIoTHubModuleClient_LL_SendEventToOutputAsync = NULL;
-              ptThisInst->tIotConfig.pfnMQTT_Protocol                                = NULL;
-            }
-            else
-            {
-              /* everything ok */
-              ptThisInst->tIotConfig.fIOTLibLoadOk = 1;
-            }
-          }
-          else
-          {
-            /* azure iot not available */
-          }
-          
-
           if (getCurrentModuleAbsolutePath(pchModulePath, pchBackendLib))
           {
             /* trying to load Paho for MQTT messaging */
@@ -812,12 +529,6 @@
     case CMD_ID_RBSTART:
     {
 
-      if (ptThisInst->tIotConfig.fIOTLibLoadOk == 1)
-      {
-        // connect only on RB_START 
-        iothub_module_init(ptThisInst);
-      }
-
       if (ptThisInst->tMqttCfg.fPahoOk == 1)
       {
         int iMqttRes = MQTTCLIENT_FAILURE;
@@ -858,12 +569,6 @@
 
     case CMD_ID_RBSTOP:
     { 
-      if (ptThisInst->tIotConfig.fIOTLibLoadOk == 1)
-      {
-        /* cleanup all IoTHub things */
-        iothub_module_deinit(ptThisInst);
-      }
-
       if (ptThisInst->tMqttCfg.fPahoOk == 1)
       {
         if (ptThisInst->tMqttCfg.pfnMQTTClient_disconnect != NULL)
@@ -1227,44 +932,6 @@
             }
           }
 
-          /* here we construct an iot message that holds a copy for the forward queue */
-          if (ptThisInst->tIotConfig.iotHubModuleClientHandle != NULL && ptMessageSinkParams->szAmqpTopic)
-          {
-            MESSAGE_INSTANCE *messageInstance = CreateMessageInstance(szMessage, ptThisInst);
-            IOTHUB_MODULE_CLIENT_LL_HANDLE iotHubModuleClientHandle = ptThisInst->tIotConfig.iotHubModuleClientHandle;
-            if (NULL == messageInstance)
-            {
-              // TODO: here should be a state callback indicating a lost connection 
-              //result = IOTHUBMESSAGE_ABANDONED; 
-            }
-            else
-            {
-
-              IOTHUB_CLIENT_RESULT clientResult = IOTHUB_CLIENT_OK;
-
-              /* verbose printf for debugging missing messages enable VERBOSE_DEBUG */
-              VERBOSE_PRINT("Sending message (%zu) to the next stage in pipeline\n", ptThisInst->tIotConfig.ullMessagesReceivedByInput1Queue);
-             
-              clientResult = ptThisInst->tIotConfig.pfnIoTHubModuleClient_LL_SendEventToOutputAsync(iotHubModuleClientHandle, messageInstance->messageHandle, ptMessageSinkParams->szAmqpTopic, SendConfirmationCallback, (void *)messageInstance);
-              if (clientResult != IOTHUB_CLIENT_OK)
-              {
-                ptThisInst->tIotConfig.pfnIoTHubMessage_Destroy(messageInstance->messageHandle);
-                free(messageInstance);
-                printf("IoTHubModuleClient_LL_SendEventToOutputAsync failed on sending msg#=%zu, err=%d\n", ptThisInst->tIotConfig.ullMessagesReceivedByInput1Queue, clientResult);
-                ullErrorCnt ++;
-                //result = IOTHUBMESSAGE_ABANDONED;
-              }
-              else
-              {
-
-                /* Successfully pushed message; nothing special to do here */
-                // result = IOTHUBMESSAGE_ACCEPTED;
-                ullSendOk++;
-              }
-            }
-            // increment 
-            ptThisInst->tIotConfig.ullMessagesReceivedByInput1Queue++;
-          }
         }
       }
 
diff -aurN a/msg_sink_mod/msg_sink_mod.h b/msg_sink_mod/msg_sink_mod.h
--- a/msg_sink_mod/msg_sink_mod.h	2019-04-25 07:47:31.000000000 +0200
+++ b/msg_sink_mod/msg_sink_mod.h	2019-04-25 07:58:47.489752163 +0200
@@ -9,17 +9,6 @@
 #include "interlock_helper.h"
 #include "MQTTClient.h"
 
-/* azure iot sdk includes */
-#include "iothub_module_client_ll.h"
-#include "iothub_client_options.h"
-#include "iothub_message.h"
-#include "azure_c_shared_utility/threadapi.h"
-#include "azure_c_shared_utility/crt_abstractions.h"
-#include "azure_c_shared_utility/platform.h"
-#include "azure_c_shared_utility/shared_util_options.h"
-#include "iothubtransportmqtt.h"
-#include "iothub.h"
-
 #ifdef __cplusplus
   extern "C" {
 #endif  /* _cplusplus */
@@ -30,20 +19,6 @@
 typedef int (*PFN_MQTTClient_destroy) (MQTTClient* handle);
 typedef int (*PFN_MQTTClient_publishMessage) (MQTTClient handle, const char* topicName, MQTTClient_message* msg, MQTTClient_deliveryToken* dt);
 
-typedef int                           (*PFN_IoTHub_Init) ();                                                                           // IoTHub_Init
-typedef void                          (*PFN_IoTHub_Deinit) ();                                                                         // IoTHub_Deinit();
-typedef IOTHUB_MESSAGE_HANDLE         (*PFN_IoTHubMessage_CreateFromString) (const char* source);                                      // IoTHubMessage_CreateFromString(szMessage);
-typedef void                          (*PFN_IoTHubMessage_Destroy) (IOTHUB_MESSAGE_HANDLE iotHubMessageHandle);                        // IoTHubMessage_Destroy(messageHandle);
-typedef IOTHUB_MODULE_CLIENT_LL_HANDLE(*PFN_IoTHubModuleClient_LL_CreateFromEnvironment) (IOTHUB_CLIENT_TRANSPORT_PROVIDER protocol);  // maybe IOTHUB_MODULE_CLIENT_LL_HANDLE  IOTHUB_CLIENT_CORE_LL_HANDLE IoTHubModuleClient_LL_CreateFromEnvironment(MQTT_Protocol);
-typedef void                          (*PFN_IoTHubModuleClient_LL_Destroy) (IOTHUB_MODULE_CLIENT_LL_HANDLE iotHubModuleClientHandle ); // IoTHubModuleClient_LL_Destroy(iotHubModuleClientHandle);
-typedef void                          (*PFN_IoTHubModuleClient_LL_DoWork) (IOTHUB_MODULE_CLIENT_LL_HANDLE iotHubModuleClientHandle);   // IoTHubModuleClient_LL_DoWork(iotHubModuleClientHandle);  
-typedef IOTHUB_CLIENT_RESULT          (*PFN_IoTHubModuleClient_LL_SendEventToOutputAsync) (IOTHUB_MODULE_CLIENT_LL_HANDLE iotHubModuleClientHandle, 
-                                                                              IOTHUB_MESSAGE_HANDLE eventMessageHandle, 
-                                                                              const char* outputName, 
-                                                                              IOTHUB_CLIENT_EVENT_CONFIRMATION_CALLBACK eventConfirmationCallback, 
-                                                                              void* userContextCallback);                             // IoTHubModuleClient_LL_SendEventToOutputAsync(iotHubModuleClientHandle, messageInstance->messageHandle, ptMessageSinkParams->szAmqpTopic, SendConfirmationCallback, (void *)messageInstance);
-typedef const TRANSPORT_PROVIDER*     (*PFN_MQTT_Protocol)(void);                                                                     // MQTT_Protocol(void)                                                                            
-
 typedef struct MESSAGE_SINK_PARAMS_Ttag
 {
   LIST_ENTRY(MESSAGE_SINK_PARAMS_Ttag) tList;
@@ -74,28 +49,10 @@
 typedef struct IOT_CONFIG_Ttag
 {
   int                            fEnabled;   /* if set to 0: no IOT messaging is used and other fields of this structure are not initialized, is set to 1: IOT messaging is used */
-  int                            fIOTLibLoadOk;
   int                            fInitialized;
-  HMODULE                        hIoTHubDll;
-  pthread_t                      hWorkerThread;
-  pthread_mutex_t                tThreadLock;
-  IOTHUB_MODULE_CLIENT_LL_HANDLE iotHubModuleClientHandle;
 
   uint64_t                       ullMessagesReceivedByInput1Queue;
 
-  /* When running on a container created by IoT Edge, we don't need a connection string.  Instead the SDK can derive
-      its connection material - including security tokens - directly from the IoTHubModuleClient_CreateFromEnvironment call. */
-
-  PFN_IoTHub_Init                                   pfnIoTHub_Init;
-  PFN_IoTHub_Deinit                                 pfnIoTHub_Deinit;
-  PFN_IoTHubMessage_CreateFromString                pfnIoTHubMessage_CreateFromString;
-  PFN_IoTHubMessage_Destroy                         pfnIoTHubMessage_Destroy;
-  PFN_IoTHubModuleClient_LL_CreateFromEnvironment   pfnIoTHubModuleClient_LL_CreateFromEnvironment;
-  PFN_IoTHubModuleClient_LL_Destroy                 pfnIoTHubModuleClient_LL_Destroy;
-  PFN_IoTHubModuleClient_LL_DoWork                  pfnIoTHubModuleClient_LL_DoWork;
-  PFN_IoTHubModuleClient_LL_SendEventToOutputAsync  pfnIoTHubModuleClient_LL_SendEventToOutputAsync;
-  PFN_MQTT_Protocol                                 pfnMQTT_Protocol;
- 
 } IOT_CONFIG_T;
 
 typedef struct MODULE_INSTANCE_Ttag
