diff -aurN a/term-utils/agetty.c b/term-utils/agetty.c
--- a/term-utils/agetty.c	2019-12-20 15:04:38.781912418 +0100
+++ b/term-utils/agetty.c	2021-09-29 08:10:20.375941518 +0200
@@ -1580,6 +1580,35 @@
 	return ret;
 }
 
+static char* read_text_file(const char *filename)
+{
+	int fd  = -1;
+	struct stat st;
+	char* ret = NULL;
+
+	fd = open(filename, O_RDONLY);
+	if(fd != -1) {
+		if (fstat(fd, &st) < 0 || st.st_size > 4 * 1024) {
+			log_warn(_("cannot open file %s"), filename);
+		} else {
+			ret = malloc(st.st_size + 1);
+			if(!ret)
+				log_err(_("failed to allocate memory: %m"));
+			else {
+				char *newline;
+
+				read(fd, ret, st.st_size);
+				if ((newline=strchr(ret, '\n')) != NULL)
+					*newline = '\0';
+				ret[st.st_size] = 0;
+			}
+		}
+		close(fd);
+	}
+
+	return ret;
+}
+
 #ifdef AGETTY_RELOAD
 static void open_netlink(void)
 {
@@ -2757,6 +2786,57 @@
 		}
 		break;
 	}
+	case 'F': /* print firmware version */
+	{
+		char* version = read_text_file("/firmware.version");
+
+		fprintf(ie->output, "%s", version);
+
+		free(version);
+		break;
+	}
+	case 'I': /* print hardware ID */
+	{
+		const int MAX_SN_LEN=12, MAX_PN_LEN=12;
+		const char* USE_CLOUD_ID="FFFFFFFFFFFF";
+		char* product_number = read_text_file("/var/platform/device_data/product_number");
+		if(strncmp(product_number, USE_CLOUD_ID, strlen(USE_CLOUD_ID)) == 0)
+		{
+			char* hardware_id=read_text_file("/mnt/backup/nvd/hardware_id");
+			fprintf(ie->output, "%s", hardware_id);
+			free(hardware_id);
+		} else {
+			char* serial_number  = read_text_file("/var/platform/device_data/serial_number");
+			const char *padding  = "000000000000";
+			int sn_padding = 0, pn_padding = 0;
+			int sn_offset = 0;
+
+			if(serial_number && strlen(serial_number) > MAX_SN_LEN)
+				sn_offset = strlen(serial_number) - MAX_SN_LEN;
+			else
+				sn_padding = MAX_SN_LEN - strlen(serial_number);
+
+			if(product_number && strlen(product_number) < MAX_PN_LEN)
+				pn_padding = MAX_PN_LEN - strlen(product_number);
+
+			fprintf(ie->output, "%*.*s%.*s-%*.*s%.*s",
+				pn_padding, pn_padding, padding, MAX_PN_LEN, product_number ? product_number : "000000000000",
+				sn_padding, sn_padding, padding, MAX_SN_LEN, serial_number  ? serial_number + sn_offset : "000000000000");
+
+			free(serial_number);
+		}
+		free(product_number);
+		break;
+	}
+	case 'P': /* print product name */
+	{
+		char* model = read_text_file("/var/platform/device_data/product_name");
+
+		fprintf(ie->output, "%s", model);
+
+		free(model);
+		break;
+	}
 	case 'S':
 	{
 		char *var = NULL, varname[64];
