diff -aurN a/setup.py b/setup.py
--- a/setup.py	2013-05-27 00:44:18.000000000 +0200
+++ b/setup.py	2019-05-15 11:24:19.856763119 +0200
@@ -7,7 +7,7 @@
 # Dirty hack to make setup.py install the iotop script to sbin/ instead of bin/
 # while still honoring the choice of installing into local/ or not.
 if hasattr(distutils_install, 'INSTALL_SCHEMES'):
-    for d in distutils_install.INSTALL_SCHEMES.itervalues():
+    for d in distutils_install.INSTALL_SCHEMES.values():
         if d.get('scripts', '').endswith('/bin'):
             d['scripts'] = d['scripts'][:-len('/bin')] + '/sbin'
 
