diff -aurN a/src/libnm-platform/nm-linux-platform.c b/src/libnm-platform/nm-linux-platform.c
--- a/src/libnm-platform/nm-linux-platform.c	2022-03-07 10:09:54.000000000 +0100
+++ b/src/libnm-platform/nm-linux-platform.c	2024-01-12 12:03:40.122319418 +0100
@@ -8138,6 +8138,9 @@
     nm_assert(NM_IN_SET(props->type, IFF_TAP, IFF_TUN));
     nm_assert(props->persist || out_fd);
 
+    if (strncmp(name,"cifx",strlen("cifx")) == 0)
+       return FALSE;
+
     fd = open("/dev/net/tun", O_RDWR | O_CLOEXEC);
     if (fd < 0)
         return FALSE;
