diff -aurN a/src/firewalld.in b/src/firewalld.in
--- a/src/firewalld.in	2022-07-26 14:42:51.151838701 +0200
+++ b/src/firewalld.in	2022-07-26 14:44:33.134379674 +0200
@@ -83,20 +83,6 @@
             log.addInfoLogging("*", log.stdout)
             log.addDebugLogging("*", log.stdout)
 
-    log_file = FileLog(config.FIREWALLD_LOGFILE, "a")
-    try:
-        log_file.open()
-    except IOError as e:
-        log.error("Failed to open log file '%s': %s", config.FIREWALLD_LOGFILE,
-                  str(e))
-    else:
-        log.addInfoLogging("*", log_file, [ log.FATAL, log.ERROR, log.WARNING,
-                                            log.TRACEBACK ])
-        log.addDebugLogging("*", log_file)
-        if args.debug:
-            log.addInfoLogging("*", log_file)
-            log.addDebugLogging("*", log_file)
-
 def startup(args):
     try:
         if not args.nofork:
