diff --git a/src/firewall-cmd.in b/src/firewall-cmd.in
index 2db183e6..6dd10c0b 100755
--- a/src/firewall-cmd.in
+++ b/src/firewall-cmd.in
@@ -37,7 +37,8 @@ from firewall import errors
 from firewall.functions import joinArgs, splitArgs, getPortRange
 from firewall.core.fw_nm import nm_is_imported, \
     nm_get_connection_of_interface, nm_get_zone_of_connection, \
-    nm_set_zone_of_connection, nm_get_interfaces_in_zone
+    nm_set_zone_of_connection, nm_get_interfaces_in_zone, \
+    nm_get_connections
 from firewall.core.io.zone import zone_reader
 from firewall.core.io.policy import policy_reader
 from firewall.core.io.service import service_reader
@@ -517,7 +518,22 @@ def try_set_zone_of_interface(_zone, interface):
                     cmd.print_msg("The interface is under control of NetworkManager, setting zone to default.")
                 else:
                     cmd.print_msg("The interface is under control of NetworkManager, setting zone to '%s'." % _zone)
+                # On active interfaces NetworkManager will send Add/Remove/ChangeZone, but for inactive interfaces, we
+                # need to do it manually nm_get_connections lists all active connections
+                old_zone = nm_get_zone_of_connection(connection)
+                connections = {}
+                connection_names = {}
+                nm_get_connections(connections, connection_names)
+
                 nm_set_zone_of_connection(_zone, connection)
+
+                if connection not in connection_names:
+                    # Update inactive connection manually
+                    if old_zone != _zone:
+                        fw.removeInterface(old_zone, interface)
+                    if _zone != "":
+                        fw.addInterface(_zone, interface)
+
                 return True
     return False
 
diff --git a/src/firewall/core/fw_nm.py b/src/firewall/core/fw_nm.py
index 0e38dd47..98665141 100644
--- a/src/firewall/core/fw_nm.py
+++ b/src/firewall/core/fw_nm.py
@@ -24,7 +24,8 @@
 __all__ = [ "check_nm_imported", "nm_is_imported",
             "nm_get_zone_of_connection", "nm_set_zone_of_connection",
             "nm_get_connections", "nm_get_connection_of_interface",
-            "nm_get_bus_name", "nm_get_dbus_interface" ]
+            "nm_get_bus_name", "nm_get_dbus_interface",
+            "nm_reset_zone_for_unassigned_interfaces" ]
 
 import gi
 from gi.repository import GLib
@@ -175,17 +176,36 @@ def nm_get_interfaces():
             if ip_iface:
                 active_interfaces.append(ip_iface)
 
+    connections = nm_get_client().get_connections()
+
+    for c in connections:
+        s_conn = c.get_setting(NM.SettingConnection)
+        c_itf = s_conn.get_interface_name()
+        if c_itf not in active_interfaces:
+            active_interfaces.append(c_itf)
+
     return active_interfaces
 
 def nm_get_interfaces_in_zone(zone):
     interfaces = []
     for interface in nm_get_interfaces():
         conn = nm_get_connection_of_interface(interface)
-        if zone == nm_get_zone_of_connection(conn):
+        if conn != None and zone == nm_get_zone_of_connection(conn):
             interfaces.append(interface)
 
     return interfaces
 
+def _nm_get_inactive_connection_of_interface(interface):
+    connections = nm_get_client().get_connections()
+
+    for c in connections:
+        s_conn = c.get_setting(NM.SettingConnection)
+        c_itf = s_conn.get_interface_name();
+        if c_itf == interface:
+            return c.get_uuid()
+
+    return None
+
 def nm_get_connection_of_interface(interface):
     """Get connection from NM that is using the interface
     @param interface name
@@ -199,7 +219,7 @@ def nm_get_connection_of_interface(interface):
 
     active_con = device.get_active_connection()
     if active_con is None:
-        return None
+        return _nm_get_inactive_connection_of_interface(interface)
 
     try:
         con = active_con.get_connection()
@@ -230,3 +250,12 @@ def nm_get_dbus_interface():
     if not _nm_imported:
         return ""
     return NM.DBUS_INTERFACE
+
+def nm_reset_zone_for_unassigned_interfaces(assigned_interfaces):
+    connections = nm_get_client().get_connections()
+
+    for c in connections:
+        s_conn = c.get_setting(NM.SettingConnection)
+        c_itf = s_conn.get_interface_name();
+        if not c_itf in assigned_interfaces:
+            nm_set_zone_of_connection(None, c.get_uuid())
diff --git a/src/firewall/server/firewalld.py b/src/firewall/server/firewalld.py
index f5f902d6..80c3199c 100644
--- a/src/firewall/server/firewalld.py
+++ b/src/firewall/server/firewalld.py
@@ -25,6 +25,7 @@ from gi.repository import GLib
 import copy
 import dbus
 import dbus.service
+import os
 
 from firewall import config
 from firewall.core.fw import Firewall
@@ -49,7 +50,9 @@ from firewall.core.io.ipset import IPSet
 from firewall.core.io.icmptype import IcmpType
 from firewall.core.io.helper import Helper
 from firewall.core.fw_nm import nm_get_connection_of_interface, \
-                                nm_set_zone_of_connection
+                                nm_set_zone_of_connection, \
+                                nm_reset_zone_for_unassigned_interfaces, \
+                                nm_get_interfaces
 from firewall.core.fw_ifcfg import ifcfg_set_zone_of_interface
 from firewall import errors
 from firewall.errors import FirewallError
@@ -443,17 +446,14 @@ class FirewallD(DbusServiceObject):
         # zones
 
         config_names = self.config.getZoneNames()
+        assigned_interfaces = []
         for name in self.fw.zone.get_zones():
             conf = self.getZoneSettings2(name)
             settings = FirewallClientZoneSettings(conf)
             changed = False
-            for interface in settings.getInterfaces():
-                if interface in self.fw._nm_assigned_interfaces:
-                    log.debug1("Zone '%s': interface binding for '%s' has been added by NM, ignoring." % (name, interface))
-                    settings.removeInterface(interface)
-                    changed = True
-            # For the remaining interfaces, attempt to let NM manage them
-            for interface in settings.getInterfaces():
+            interfaces_copy = settings.getInterfaces().copy()
+            for interface in interfaces_copy:
+                assigned_interfaces.append(interface)
                 try:
                     connection = nm_get_connection_of_interface(interface)
                     if connection and nm_set_zone_of_connection(name, connection):
@@ -499,6 +499,7 @@ class FirewallD(DbusServiceObject):
                     (name, e))
                 error = True
 
+        nm_reset_zone_for_unassigned_interfaces(assigned_interfaces)
         # helpers
 
         config_names = self.config.getHelperNames()
@@ -1289,6 +1290,15 @@ class FirewallD(DbusServiceObject):
         """
         zone = dbus_to_python(zone, str)
         interface = dbus_to_python(interface, str)
+
+        sender_cmd = command_of_sender(dbus.SystemBus(), sender)
+        sender_cmd = sender_cmd.split(' ')[0]
+        sender_cmd = os.path.basename(sender_cmd)
+        # Deny removal of networkmanager interfaces, as we want to keep interfaces that are down
+        if sender_cmd == "NetworkManager":
+            log.debug1("zone.removeInterface('%s', '%s') - Skipping removal of NetworkManager interface" % (zone, interface))
+            return zone
+
         log.debug1("zone.removeInterface('%s', '%s')" % (zone, interface))
         self.accessCheck(sender)
         _zone = self.fw.zone.remove_interface(zone, interface)
