diff -aurN a/Makefile b/Makefile
--- a/Makefile	2021-04-13 19:30:13.748440002 +0200
+++ b/Makefile	2021-04-13 19:30:30.696755488 +0200
@@ -294,7 +294,6 @@
 INSTALL_DATA = $(INSTALL) -m 644
 
 # Do not invoke directly; this is invoked by `rpmbuild` or `dpkg-buildpackage`. Use `make centos` or `make deb` instead.
-install-common: default
 install-common:
 	# Ref: https://www.gnu.org/software/make/manual/html_node/DESTDIR.html
 
