diff -aurN a/Makefile b/Makefile
--- a/Makefile	2024-03-18 10:04:10.691907657 +0100
+++ b/Makefile	2024-03-18 10:07:16.452540636 +0100
@@ -56,20 +56,6 @@
 	CARGO_PROFILE_DIRECTORY = release
 endif
 
-ifneq (,$(filter $(ARCH), arm32v7 armhf))
-	CARGO_TARGET = armv7-unknown-linux-gnueabihf
-    CROSS_HOST_TRIPLE = arm-linux-gnueabihf
-	DPKG_ARCH_FLAGS = --host-arch armhf
-else ifneq (,$(filter $(ARCH), aarch64 arm64))
-	CARGO_TARGET = aarch64-unknown-linux-gnu
-    CROSS_HOST_TRIPLE = aarch64-linux-gnu
-	DPKG_ARCH_FLAGS = --host-arch arm64 --host-type aarch64-linux-gnu --target-type aarch64-linux-gnu
-else
-	CARGO_TARGET = x86_64-unknown-linux-gnu
-    CROSS_HOST_TRIPLE = x86_64-linux-gnu
-	DPKG_ARCH_FLAGS =
-endif
-
 CARGO_OUTPUT_ABSPATH = $(abspath ./target/$(CARGO_TARGET)/$(CARGO_PROFILE_DIRECTORY))
 VENDOR_PREFIX = $(CARGO_OUTPUT_ABSPATH)/fakeroot
 VENDOR_PKGCONFIG = $(VENDOR_PREFIX)$(AZIOT_PRIVATE_LIBRARIES)/pkgconfig
@@ -81,6 +67,8 @@
 		USER_AZIOTTPM="$(USER_AZIOTTPM)" \
 		SOCKET_DIR="$(SOCKET_DIR)" cargo
 
+CARGO_TARGET ?= x86_64-unknown-linux-gnu
+
 # Some of the targets use bash-isms like `set -o pipefail`
 SHELL = /bin/bash
 
