diff --git a/Makefile b/Makefile
index 9141611..f70dadc 100644
--- a/Makefile
+++ b/Makefile
@@ -1,4 +1,5 @@
 CARGO = cargo
+CARGO_TARGET ?= x86_64-unknown-linux-gnu
 
 # Installed via `cargo install`. See /ci/install-build-deps.sh for exact command and versions.
 BINDGEN = bindgen
@@ -32,18 +33,6 @@ else
 	CARGO_PROFILE_DIRECTORY = release
 endif
 
-ifeq ($(ARCH), arm32v7)
-	CARGO_TARGET = armv7-unknown-linux-gnueabihf
-	DPKG_ARCH_FLAGS = --host-arch armhf
-else ifeq ($(ARCH), aarch64)
-	CARGO_TARGET = aarch64-unknown-linux-gnu
-	DPKG_ARCH_FLAGS = --host-arch arm64 --host-type aarch64-linux-gnu --target-type aarch64-linux-gnu
-else
-	CARGO_TARGET = x86_64-unknown-linux-gnu
-	DPKG_ARCH_FLAGS =
-endif
-
-
 # Some of the targets use bash-isms like `set -o pipefail`
 SHELL := /bin/bash
 
