diff -aurN a/Makefile b/Makefile
--- a/Makefile	2022-01-28 08:26:37.718246931 +0100
+++ b/Makefile	2022-01-28 08:31:09.644953689 +0100
@@ -1,4 +1,5 @@
 CARGO = cargo
+CARGO_TARGET ?= x86_64-unknown-linux-gnu
 
 # Installed via `cargo install`. See /ci/install-build-deps.sh for exact command and versions.
 BINDGEN = bindgen
@@ -34,17 +35,6 @@
 	CARGO_PROFILE_DIRECTORY = release
 endif
 
-ifeq ($(ARCH), arm32v7)
-	CARGO_TARGET = armv7-unknown-linux-gnueabihf
-	DPKG_ARCH_FLAGS = --host-arch armhf
-else ifeq ($(ARCH), aarch64)
-	CARGO_TARGET = aarch64-unknown-linux-gnu
-	DPKG_ARCH_FLAGS = --host-arch arm64 --host-type aarch64-linux-gnu --target-type aarch64-linux-gnu
-else
-	CARGO_TARGET = x86_64-unknown-linux-gnu
-	DPKG_ARCH_FLAGS =
-endif
-
 # Some of the targets use bash-isms like `set -o pipefail`
 SHELL := /bin/bash
 
