diff -aurN a/contrib/third-party-notices.sh b/contrib/third-party-notices.sh
--- a/contrib/third-party-notices.sh	2021-04-15 10:51:10.582872468 +0200
+++ b/contrib/third-party-notices.sh	2021-04-15 10:58:53.356118613 +0200
@@ -5,25 +5,6 @@
 # This script identifies all the third-party dependencies of the crates in the workspace.
 # At a high level, `cargo metadata --format-version 1` gives this information in JSON.
 
-case "$ARCH" in
-    'amd64')
-        platform='x86_64-unknown-linux-gnu'
-        ;;
-
-    'arm32v7')
-        platform='armv7-unknown-linux-gnueabihf'
-        ;;
-
-    'aarch64')
-        platform='aarch64-unknown-linux-gnu'
-        ;;
-
-    *)
-        echo "Unsupported ARCH $ARCH" >&2
-        exit 1
-        ;;
-esac
-
 cargo_metadata="$(
     cargo metadata --format-version 1 --filter-platform "$platform" |
         jq -r '.resolve.nodes'
