From d2c47a973f354ffd505bb4e809c59e57b543726d Mon Sep 17 00:00:00 2001
From: Chen Qi <Qi.Chen@windriver.com>
Date: Tue, 6 Aug 2019 19:01:45 +0800
Subject: [PATCH] Makefile: respect GOBUILDFLAGS for runc and remove recvtty
 from static

Signed-off-by: Chen Qi <Qi.Chen@windriver.com>
---
 Makefile | 1 +
 1 file changed, 1 insertion(+), 0 deletions(-)

diff -aurN a/src/import/Makefile b/src/import/Makefile
--- a/src/import/Makefile	2022-05-09 07:32:54.499134853 +0200
+++ b/src/import/Makefile	2022-05-09 07:35:52.627639716 +0200
@@ -25,6 +25,7 @@
 	endif
 endif
 GO_BUILD := $(GO) build -trimpath $(GO_BUILDMODE) \
+	$(GOBUILDFLAGS) \
 	$(EXTRA_FLAGS) -tags "$(BUILDTAGS)" \
 	-ldflags "$(LDFLAGS_COMMON) $(EXTRA_LDFLAGS)"
 
