--- a/netx_tap/netx_tap.c	2024-04-02 18:20:33.018535588 +0200
+++ b/netx_tap/netx_tap.c	2024-04-02 18:20:59.127080730 +0200
@@ -79,7 +79,7 @@
 
 static TAILQ_HEAD(, NETX_ETH_DEV_Ttag) s_DeviceList = TAILQ_HEAD_INITIALIZER( s_DeviceList);
 
-static int32_t cifxeth_search_eth_channel   ( char* szDeviceName, uint32_t ulSearchIdx, uint32_t* pulChannelNumber);
+       int32_t cifxeth_search_eth_channel   ( char* szDeviceName, uint32_t ulSearchIdx, uint32_t* pulChannelNumber);
 static int     cifxeth_allocate_tap         ( NETX_ETH_DEV_T* internal_dev, char* prefix);
 static void    cifxeth_free_tap             ( NETX_ETH_DEV_T* internal_dev, char* name);
 static void    cifxeth_delete_device        ( NETX_ETH_DEV_T* internal_dev);
