--- a/user_linux.c	2024-04-02 18:16:37.693600868 +0200
+++ b/user_linux.c	2024-04-02 18:18:00.743347288 +0200
@@ -42,15 +42,15 @@
 #endif
 
 uint8_t severity_mapping[] = {
-  0xff, /* not used */
-  1,    /* TRACE_LEVEL_DEBUG */
-  2,    /* TRACE_LEVEL_INFO */
+  7,    /* header */
+  7,    /* TRACE_LEVEL_DEBUG */
+  6,    /* TRACE_LEVEL_INFO */
   0xff, /* not used */
   4,    /* TRACE_LEVEL_WARNING */
   0xff, /* not used */
   0xff, /* not used */
   0xff, /* not used */
-  8,    /* TRACE_LEVEL_ERROR */
+  3,    /* TRACE_LEVEL_ERROR */
 };
 
 /*****************************************************************************/
